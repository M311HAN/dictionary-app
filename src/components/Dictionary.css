/* Dictionary.css */
body,
html {
  margin: 0;
  padding: 0;
  height: 100%;
  width: 100%;
  font-family: Arial, sans-serif;
  background-color: #f4f4f4; /* Light grey background for contrast */
}

.container {
  padding: 20px;
  text-align: center;
  max-width: 800px;
  margin: 0 auto;
}

h1,
h2 {
  font-family: "Cinzel", serif;
  color: #d4af37; /* Gold color */
  text-shadow: 1px 1px 1px #000; /* Black shadow */
  margin-bottom: 20px;
}

h1 {
  font-size: 3em;
}

h2 {
  font-size: 2em;
  margin-top: 20px;
}

form {
  display: flex;
  justify-content: center;
  margin-bottom: 20px;
}

input[type="text"] {
  padding: 10px;
  font-size: 1em;
  border: none;
  border-radius: 5px 0 0 5px;
  outline: none;
  box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.2); /* Light shadow */
}

button {
  padding: 10px;
  font-size: 1em;
  border: none;
  border-radius: 0 5px 5px 0;
  background-color: #d4af37; /* Gold color */
  color: white;
  cursor: pointer;
  font-family: "Cinzel", serif;
  text-shadow: 1px 1px 2px #000; /* Black shadow */
  box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.2); /* Light shadow */
}

button:hover {
  background-color: #b08930; /* Darker gold */
}

p {
  margin: 10px 0;
}

.definition-container,
.example-container {
  padding: 0 20px; /* Add padding to the left and right */
  text-align: center; /* Keep text centered */
}

/* Media Queries for Responsive Design */
@media (max-width: 600px) {
  h1 {
    font-size: 2.5em;
  }

  h2 {
    font-size: 1.5em;
    text-shadow: 0.5px 0.5px 1px #000; /* Smaller shadow on mobile */
  }

  .container {
    padding: 10px;
  }

  .definition-container,
  .example-container {
    padding: 0 10px; /* Reduce padding on smaller screens */
  }

  input[type="text"],
  button {
    font-size: 0.9em;
    padding: 8px;
  }
}
