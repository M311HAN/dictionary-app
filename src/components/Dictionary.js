import React, { useState } from "react";
import axios from "axios";
import "./Dictionary.css";

// Dictionary component that fetches and displays definitions and examples of words
const Dictionary = () => {
  // State variables to hold the word, definition, example, and error message
  const [word, setWord] = useState("");
  const [definition, setDefinition] = useState("");
  const [example, setExample] = useState("");
  const [error, setError] = useState("");

  // Function to fetch the definition and example from the Merriam-Webster API
  const fetchDefinition = async () => {
    try {
      // Make an HTTP GET request to the Merriam-Webster API
      const response = await axios.get(
        `https://www.dictionaryapi.com/api/v3/references/learners/json/${word}?key=${process.env.REACT_APP_DICTIONARY_API_KEY}`
      );
      const data = response.data;

      // Check if the response contains valid data
      if (data.length > 0 && typeof data[0] === "object") {
        // Set the definition from the API response
        setDefinition(
          data[0].shortdef ? data[0].shortdef[0] : "No definition found"
        );

        // Extract the example sentence if it exists
        const exampleData = data[0]?.def?.[0]?.sseq?.[0]?.[0]?.[1]?.dt?.find(
          (dtItem) => Array.isArray(dtItem) && dtItem[0] === "vis"
        );
        const exampleSentence = exampleData
          ? exampleData[1][0].t
          : "No example found";

        // Format the example sentence to replace {it} tags with <i> tags for italicization
        const formattedExample = exampleSentence
          .replace(/{it}/g, "<i>")
          .replace(/{\/it}/g, "</i>");

        // Set the example sentence in the state
        setExample(formattedExample || "No example found");
      } else {
        // Set default messages if no valid data is found
        setDefinition("No definition found");
        setExample("No example found");
      }
    } catch (err) {
      console.error(err); // Log the error
      setError("Failed to fetch definition");
    }
  };

  // Handle form submission
  const handleSubmit = (e) => {
    e.preventDefault();
    fetchDefinition();
  };

  return (
    <div className="container">
      <h1>Dictionary App</h1>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          value={word}
          onChange={(e) => setWord(e.target.value)}
          placeholder="Enter a word"
        />
        <button type="submit">Get Definition</button>
      </form>
      {error && <p>{error}</p>}
      {definition && (
        <div className="definition-container">
          <h2>Definition:</h2>
          <p>{definition}</p>
        </div>
      )}
      {example && (
        <div className="example-container">
          <h2>Example:</h2>
          <p dangerouslySetInnerHTML={{ __html: example }} />
        </div>
      )}
    </div>
  );
};

export default Dictionary;
