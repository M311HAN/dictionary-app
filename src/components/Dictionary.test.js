import React from "react";
import renderer from "react-test-renderer";
import Dictionary from "./Dictionary";

// Snapshot test for the Dictionary component
// This test ensures that the Dictionary component renders correctly
// and that the rendered output matches the stored snapshot.

// Note: The strikethrough on `renderer.create` indicates that it might be
// deprecated or that there is a newer preferred method. However, as of
// react-test-renderer version 18.3.1, this method is still widely used
// and functions correctly for snapshot testing.

test("Dictionary component renders correctly", () => {
  // Create a snapshot of the Dictionary component
  const tree = renderer.create(<Dictionary />).toJSON();

  // Compare the rendered component to the stored snapshot
  // If the rendered output changes, the test will fail
  expect(tree).toMatchSnapshot();
});
