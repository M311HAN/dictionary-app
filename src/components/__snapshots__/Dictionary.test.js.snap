// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dictionary component renders correctly 1`] = `
<div
  className="container"
>
  <h1>
    Dictionary App
  </h1>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Enter a word"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Get Definition
    </button>
  </form>
</div>
`;
